Core/Src/main.c

```diff
@@ -87,14 +103,45 @@ int main(void)
   /* Initialize all configured peripherals */
   MX_GPIO_Init();
   MX_USART2_UART_Init();
+  MX_SPI1_Init();
+  MX_TIM1_Init();
   /* USER CODE BEGIN 2 */
 
+  // initialize with chip select = 1
+  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+  // start timer
+  HAL_TIM_PWM_Start(&htim1, TIM_CHANNEL_1);
+
   /* USER CODE END 2 */
 
   /* Infinite loop */
   /* USER CODE BEGIN WHILE */
   while (1)
   {
+	  // chip select = 0 to begin communication
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_RESET);
+
+	  // receive data from ADC
+	  HAL_SPI_TransmitReceive(&hspi1, (uint8_t*) TRANSMIT_DATA, receive_data, size, TIMEOUT);
+
+	  // chip select = 1 to end communication
+	  HAL_GPIO_WritePin(GPIOB, GPIO_PIN_8, GPIO_PIN_SET);
+
+	  // get 10 bits of data
+	  adc = ((uint16_t) receive_data[1] << 8 | (uint16_t) receive_data[2]);
+	  // bit masking
+	  adc = adc & 0x03FF;
+
+	  // convert adc to counts
+	  compare_data = (adc/ADC_MAX)*(COUNTER_PERIOD)*(MIN_DUTY_CYCLE) + (COUNTER_PERIOD)*(MIN_DUTY_CYCLE);
```

Comment:
This line might cause a loss of precision due to integer division

Reply:
Also not sure why everything is multiplied by the MIN_DUTY_CYCLE, that is just the offset away from 0.
I would recommend splitting up these math operations to different lines and commenting what they are doing.

Reply:
I added comments to explain my calculations, if it is still confusing then I can separate to different lines